### ISSUE TITLE
[ENH] Enhancements for torch based classifiers and regressors

### ISSUE THREAD

**Issue by RecreationalMath** [alan-turing-institute/sktime#8928]
### Is your feature request related to a problem?

Not related to an issue. More of a code-refactor & coming up with a framework for PyTorch-based classifiers and regressors.

---

### Description
Currently, the deep learning estimators are primarily implemented in TensorFlow and have a very clean and concise design/implementation. As part of #8699  we're implementing these deep learning estimators in PyTorch. 
This issue is a placeholder for all the design discussions regarding PyTorch interfaces.

1. Dataset and dataloader can be abstracted out of the base PyTorch classifier in `_pytorch.py` to reduce the code duplication in the classifier and regressor classes. Doing so will also consolidate the logic and make it less error-prone and more general implementations of the common functions can be achieved.
2. Optimizer and criterion validation can be abstracted to a separate helper/utility file, so that base classifier and base regressor can reuse it from a single place, it will be less error-prone to future changes and easier to maintain.
3. Directory structure & naming conventions for estimator & network files can be borrowed from #8842 

---

### Additional context
Some of the older discussions on the same topic
- #6641
- #6790 
- #7598

**Comment by RecreationalMath** [comment]
[Discord discussion thread](https://discord.com/channels/1075852648688930887/1424918579819581591) on handling soft dependencies of torch and the best way to localise them at the module level.

**Comment by RecreationalMath** [comment]
A good point made by @achieveordie [here](https://github.com/sktime/sktime/pull/8842#discussion_r2415849238) about optimizer and criterion importing and keeping the size of the estimator class down.

**Comment by RecreationalMath** [comment]
Each PyTorch-based estimator should expose both activations (hidden & output-layer) to the user. Refer to the discussion here: #8880

**Comment by RecreationalMath** [comment]
> Each PyTorch-based estimator should expose both activations (hidden & output-layer) to the user. Refer to the discussion here: [#8880](https://github.com/sktime/sktime/pull/8880)

When doing so for classifiers, 
- Certain combinations are not numerically stable and have stable functionally equivalent alternatives, which are preferred. 
- Additionally, certain loss functions internally implement the last layer activation; in such cases, activation needs to be enforced to be None. 
Refer discussion in this [comment](https://github.com/sktime/sktime/pull/8842#discussion_r2404383748)

Separate tests (including negative tests) should be written for these cases to enforce valid inputs. `get_test_params` can not be used for this. `get_test_params` expects valid/correct/non-error-raising inputs, as it is used only for framework testing. Refer [discord comment here](https://discord.com/channels/1075852648688930887/1080978709412069426/1427474075546026126)

**Comment by RecreationalMath** [comment]
Metrics for evaluation. 
- Do we need to add support for multiple metrics in PyTorch-based estimators, similar to how it can be passed in `model.compile` as an argument in TensorFlow-based estimators?
- While PyTorch doesn't have something in-built akin to TensorFlow, the `torchmetrics` library can be utilised here, though it will add another dependency.
- Currently, `Accuracy` is supported in PyTorch-based estimators, which is the default setting of metrics in TensorFlow-based estimators.

**Comment by RecreationalMath** [comment]
Notes on testing of base classes for PyTorch-based classifiers and regressors.
- Before #8842, there were two PyTorch-based classifiers implemented in sktime, gru.py and mvts_transformer.py. Both of these used the `BaseDeepClassifierTorch` class implemented in `_pytorch.py`. I've noticed that this `BaseDeepClassifierTorch` class was not added to the `sktime/classification/tests/test_base.py`, whereas the TensorFlow-based `BaseDeepClassifier` class is added. It seems like it's an oversight and not a deliberate choice. (**Need inputs from maintainers on this**)
- Shouldn't the `BaseDeepClassifierTorch` and eventually (once implemented) `BaseDeepRegressorTorch` both be added to their respective `test_base.py`?
- Eventually, base classes for PyTorch-based estimators will be moved from `_pytorch.py` to the new directory structure and a new file, as suggested in #8842.
- Instead of adding the base class implemented in `_pytorch.py`, once the move is completed, the new base classes should be added to the respective `test_base.py`.

**Comment by RecreationalMath** [comment]
Documentation notes:

- Each newly implemented PyTorch-based classifier and regressor should be added to its respective api_reference .rst file
- Also, the BaseDeepClassifierTorch & BaseDeepRegressorTorch should be added to the respective api_reference/ .rst file. 
- Similar to the above comment, while BaseDeepClassifierTorch class has been part of sktime for a while, it hasn't been added to the .rst file.

**Comment by RecreationalMath** [comment]
Is there a checklist of tasks to be referred to when implementing a new estimator?
The above two comments and the lack of coherence in the parameter `activation`, pointed out in #8880, could have been prevented if such a checklist existed - just a thought.

Update: There is a rather small checklist in the docstring of the default PR text. Maybe it can be enhanced or moved somewhere else.

**Comment by RecreationalMath** [comment]
If the GPU-caused speedups are not big enough, transferring the data/results from CPU to GPU and back can be time-consuming and lead to worse performance. If/when to move to GPU processing needs to be discussed. Point made by @agobbifbk  [here](https://github.com/sktime/sktime/pull/8842#issuecomment-3409101411)

**Comment by RecreationalMath** [comment]
Support for callbacks to be added to PyTorch-based estimators. Point brought up by @agobbifbk refer [comment here](https://github.com/sktime/sktime/pull/8842#issuecomment-3409609035)
Callbacks are usually used for
1. Learning Rate Schedulers
2. Early Stopping
3. Model Checkpointing
4. Logging & Monitoring

PyTorch, being a low-level API, only supports Learning Rate Schedulers. All others are usually used from a higher-level library, such as Lightning or Ignite. Adding support for all these callbacks will introduce newer dependencies. 
[Common callbacks](https://lightning.ai/docs/pytorch/stable/api_references.html#callbacks) implemented in Lightning

**Comment by fkiraly** [comment]
amazing!

To answer some questions from my perspective - this is opinion based ofc.

> I've noticed that this `BaseDeepClassifierTorch` class was not added to the `sktime/classification/tests/test_base.py`, whereas the TensorFlow-based `BaseDeepClassifier` class is added. It seems like it's an oversight and not a deliberate choice. (**Need inputs from maintainers on this**)

I think it is simply due to the fact that we have not thought about consistent base class architecture and these two efforts were separate.

> Do we need to add support for multiple metrics in PyTorch-based estimators, similar to how it can be passed in `model.compile` as an argument in TensorFlow-based estimators?

Not necessarily for the start imo - is this not a native feature of `torch`?

> While PyTorch doesn't have something in-built akin to TensorFlow, the `torchmetrics` library can be utilised here, though it will add another dependency.

It could be a soft dependency and would be entirely optional if we allow passing metrics as objects rather than strings.

> Is there a checklist of tasks to be referred to when implementing a new estimator?

There is an estimator implementation guide: https://www.sktime.net/en/latest/developer_guide/add_estimators.html

> If/when to move to GPU processing needs to be discussed

I would say - up to the user? That is, selectable by a parameter.

**Comment by RecreationalMath** [comment]
@phoeenniixx brought up dependencies being tied to a specific Python version in [this comment](https://github.com/sktime/sktime/pull/9013#discussion_r2482406888). 
To be checked: torch compatibility vis-à-vis sktime-supported Python versions on all supported OS platforms.

**Comment by fkiraly** [comment]
the tags currently do not support package dependencies based on python version, but this is something we could look into adding if it gets necessary?

**Comment by RecreationalMath** [comment]
> the tags currently do not support package dependencies based on python version, but this is something we could look into adding if it gets necessary?

Interesting! Then what does `"python_version":">=3.12"` as a tag in an estimator (let's say est_X) signify? 
I thought it meant est_X requires python_version to be >=3.12 because an imported module in est_X requires that python_version.

**Comment by RecreationalMath** [comment]
Activation functions to be used are distinguished based on whether they are passed for hidden layer activations or the final/output layer activation by having two parameters, `activation_hidden` and `activation`. Reference: #8880 and #8889 

Does something similar need to be done for using the bias term as well?
Currently, in TensorFlow-based MLP and RNN classifiers, 
- only the final/output layer bias is exposed to the user, 
- hidden-layer bias is permanently set to False (excluding the bias term to be used/learned). This is non-standard practice. This would lead to the network not learning shifts when the decision boundary does not pass through the origin, and forcing all outputs through the origin.

Current implementation of PyTorch-based RNN classifiers and regressors exposes both the hidden and final/output layer bias choice. But via a single parameter. Which I feel is at least better than always setting hidden-layer biases to False, but whether to separate/stop the forced overloading of the single bias parameter or not needs to be discussed and investigated. 

The only scenario I can think of in which `bias` term needs to be omitted is when using BatchNormalization, as BatchNorm would have its own learnable/fitted shifting parameter.

Once it is discussed and agreed upon, the behaviour should be consistent across all Deep Learning estimators, whether TensorFlow-based or PyTorch-based.

@phoeenniixx @noxthot Would appreciate your thoughts on this.

**Comment by RecreationalMath** [comment]
Perhaps torch should be added to [`project.toml`](https://github.com/sktime/sktime/blob/3e3d7683e49441bb38c90a6db7edc8cefc24dedd/pyproject.toml) (under all_extras) as TensorFlow is added. 
It would enable us to manage PyTorch and the corresponding dependent Python version as well.

**Comment by RecreationalMath** [comment]
Proposal to add PyTorch-related common functionality to a base class under `sktime.base._panel`. Check [this comment](https://github.com/sktime/sktime/pull/9013#discussion_r2483382941)

**Comment by RecreationalMath** [comment]
In TensorFlow implementations of estimators, the model is essentially split into input layer and output layer. 
- What is the reasoning behind it? 
- Is this pattern borrowed from scikit-learn or sktime specific? 
- Is there any other reason for doing so apart from re-usability of the same network in classifiiers, regressors? 
- Do these networks get used or plan to be used anywhere else in the future?
- The naming convention - `input_layer` usually refers to the first layer in a deep neural network, but here it encapsulates what is traditionally known as input_layer and hidden_layers. Also, when reading in conjuction with `output_layer` it feels like there is another component `hidden_layer` missing in the structure. What is the motivation behind this naming convention, can we modify it?

**Comment by noxthot** [comment]
> [@phoeenniixx](https://github.com/phoeenniixx) [@noxthot](https://github.com/noxthot) Would appreciate your thoughts on this.

I am really not much into the `sktime` specifics. But my gut feeling is, that the deep learning part would profit from refactoring and reconsidering the general architecture (this issue is a great starting point for this).

There is also a migration issue (from `torch` to `tensorflow`), so I am not sure about how much time should be invested on the `tensorflow` part.

**Comment by srupat** [comment]
@RecreationalMath I would like to work on some of these enhancements. Do you have any specific/important ones in mind that I can pick up and also nobody else is currently working on them?

**Comment by RecreationalMath** [comment]
> [@RecreationalMath](https://github.com/RecreationalMath) I would like to work on some of these enhancements. Do you have any specific/important ones in mind that I can pick up and also nobody else is currently working on them?

I think I already responded to this on Discord. But for the benefit of anyone looking at this: Any enhancement mentioned in the comments above (or your new ideas) which do not have a corresponding sub-issue created for it are open and welcome to be worked upon.

**Comment by varun-kht** [comment]
hi @RecreationalMath i am looking forward to contribute to this issue but almost all issues and sub issue have an active pr . If their is something unattended could you please mention it or assign it to me ?
Thank you!

**Comment by Spandan-Mishra** [comment]
Hey @RecreationalMath @fkiraly . I came across [this discussion](https://github.com/sktime/sktime/issues/8928#issuecomment-3411321011) from a few months back regarding the base test suites not having the class `BaseDeepClassifierTorch` and `BaseDeepRegressorTorch`. As discussed above we can create a separate base class architecture for Torch based classes or integrate the these classes in the existing test files. If this issue is still relevant I would love to get your inputs and work on this issue.

**Comment by nonehxrry** [comment]
Hi @RecreationalMath! I'm Harjit, interested in contributing for GSoC 2026.
I'd like to work on #9329 (Add support for different evaluation metrics in torch models). I'm comfortable with PyTorch and torchmetrics. Is this still open or is someone already working on it?
Thanks!

**Comment by Me-Priyank** [comment]
@RecreationalMath @fkiraly One design point that I don't think is on the list above: parameter naming between each TF class and its torch twin.

The torch implementations renamed several constructor parameters. Comparing `__init__` signatures across the 11 pairs in `classification/deep_learning`:

| TF | torch | pairs affected |
|---|---|---|
| `n_epochs` | `num_epochs` | 10 of 11 |
| `loss` | `criterion` | 10 of 11 |
| `kernel_size` | `kernel_sizes` | cnn, macnn, mcdcnn |
| `use_bias` | `bias` | mlp, rnn |

This matters because every public unsuffixed name still resolves to the TF class (`CNNClassifier` is `_cnn_tf`, torch is reachable only as `CNNClassifierTorch`), so pointing the public names at torch later would break anyone passing `n_epochs=`.

There is also one intra-torch outlier: `MCDCNNClassifierTorch` takes `n_epochs` and `optim`/`optim_kwargs` where the base class and the other 11 torch classes use `num_epochs` and `optimizer`/`optimizer_kwargs`. It maps them through in `super().__init__`, so it works, but it is inconsistent with its own parent.

Is harmonising these in scope here, or better as its own issue once #11171 settles? I'd like to put together the full signature diff across classification and regression if that is useful.

**Comment by srupat** [comment]
@Me-Priyank Good point! I would prefer renaming the TF params according to the torch ones, however @fkiraly can make the final call. 

As for `MCDCNNClassifierTorch`, you are right, and a bugfix PR would be appreciated for the same!
Could you also find out more discrepancies between the tf and the torch estimators? Thanks!
